### PR TITLE
Refactor the componentDidUpdate method for ReportActionCompose

### PR DESCRIPTION
**Pull request by mountiny**
### Details

To resolve a StagingDeployBlocker, we have introduced some changes which need some refactoring. How to refactor has been discussed and decided in [this thread here](https://github.com/Expensify/App/pull/4761#discussion_r693077028). Look there for more details.

### Fixed Issues

❌ 

### Tests

Just repeat the `action performed` steps from this issue https://github.com/Expensify/App/issues/4756 and make sure it all works as expected.

### QA Steps
1. Navigate to a conversation
1. Write something in compose box
1. Use ctrl+k command
1. Select another conversation
1. Verify that nothing is displayed in compose box if the selected conversation didn't have a draft
1. Use ctrl+k command to switch back to the first conversation.
1. Verify that the draft message in the compose box is shown.

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### CODE REVIEW

**Comment by mountiny** [comment]
@marcaaron My theory based on what I have observed when testing it is that the `reportID` is fetched from `route` [here](https://github.com/Expensify/App/blob/5acd2363000993f5bc946114adf83ef099e870a0/src/pages/home/ReportScreen.js#L155) and passed down to the `ReportActionCompose`. However, the `report` which holds the comment we need is not updated at that same time as the route changes. In console I could see ~2-3 calls of `componentDidUpdate` in `ReportActionCompose` with the previous `report` while the `reportID` has already been updated.

Isn't it that the `reportID` fetched from `route` is from the `Navigation` whereas report is from Onyx which can take slightly longer to switch?

**Comment by marcaaron** [comment]
>Isn't it that the reportID fetched from route is from the Navigation whereas report is from Onyx which can take slightly longer to switch?

Yes that makes sense to me. It should be like 1 tick behind because Onyx is asynchronous. Is that what you're asking?

File: src/pages/home/report/ReportActionCompose.js [hunk 1]
```diff
@@ -189,10 +189,10 @@ class ReportActionCompose extends React.Component {
             this.focus();
         }
 
-        // If we switch from a sidebar, the component does not mount again
-        // so we need to update the comment manually.
-        if (prevProps.comment !== this.props.comment) {
-            this.textInput.setNativeProps({text: this.props.comment});
+        // If we switch the reports, make sure to update the composer comment
+        const reportChanged = this.props.report.reportID !== prevProps.report.reportID;
+        if (reportChanged) {
```
Comment (marcaaron) [inline src/pages/home/report/ReportActionCompose.js:194]: This should use an early return like

```js
if (!reportChanged) {
    return;
}

this.updateComment(this.props.comment);
```

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionCompose.js (hunk 1 above)
Comment (marcaaron) [reply src/pages/home/report/ReportActionCompose.js:194]: Also `reportChanged` does not need a variable just put the comparison into the if case?

**Review by bondydaa — APPROVED** [review]

**Review by marcaaron — APPROVED** [review]

**Comment by isagoico** [comment]
@mountiny @marcaaron Are we good to check this PR off the list? (I'm guessing it's NoQA but wanted to double check)

**Comment by roryabraham** [comment]
@isagoico I filled in the QA steps here. Can you please perform QA on this?

**Comment by isagoico** [comment]
Just tested it and it was a pass 🎉 checking it off.

**Pull request merged**